pkg/metrics/context_metrics_flusher.go

```diff
@@ -0,0 +1,62 @@
+// Unless explicitly stated otherwise all files in this repository are licensed
+// under the Apache License Version 2.0.
+// This product includes software developed at Datadog (https://www.datadoghq.com/).
+// Copyright 2016-present Datadog, Inc.
+
+package metrics
+
+import (
+	"github.com/DataDog/datadog-agent/pkg/aggregator/ckey"
+)
+
+type timestampedContextMetrics struct {
+	bucketTimestamp float64
+	contextMetrics  ContextMetrics
+}
+
+// ContextMetricsFlusher flushes several ContextMetrics
+type ContextMetricsFlusher struct {
+	metrics []timestampedContextMetrics
+}
+
+// NewContextMetricsFlusher creates a new instance of ContextMetricsFlusher
+func NewContextMetricsFlusher() *ContextMetricsFlusher {
+	return &ContextMetricsFlusher{}
+}
+
+// Append appends a new contextMetrics
+func (f *ContextMetricsFlusher) Append(bucketTimestamp float64, contextMetrics ContextMetrics) {
+	f.metrics = append(f.metrics, timestampedContextMetrics{
+		bucketTimestamp: bucketTimestamp,
+		contextMetrics:  contextMetrics,
+	})
+}
+
+// FlushAndClear flushes contextMetrics to series and clear contextMetrics collection.
+// For each contextKey, FlushAndClear flushes every metrics (Same as ContextMetrics.Flush) and call
+// the callback with all series whose key context is the same.
+// If there are 3 context keys, the callback is called 3 times.
+// Note: The slice []*Serie in callback is reused.
+func (f *ContextMetricsFlusher) FlushAndClear(callback func([]*Serie)) map[ckey.ContextKey][]error {
+	errors := make(map[ckey.ContextKey][]error)
+	var series []*Serie
+
+	var contextMetricsCollection []ContextMetrics
+	for _, m := range f.metrics {
+		contextMetricsCollection = append(contextMetricsCollection, m.contextMetrics)
+	}
+	mergeContextMetrics(
```

Comment:
Having that much indirection (`mergeContextMetrics` receiving two func as parameters, one being here a callback itself passed as a parameter in `FlushAndClear`) makes the code extremely confusing and painful to read / follow.

Reply:
This code was updated. I have kept the callbacks in order to fix https://github.com/DataDog/datadog-agent/pull/10046#discussion_r760589164.
I have also updated the name of the function which I hope make more sense.